CloudWatch: Remove core imports from CloudWatchRequest

kevinwcyu opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

Removes app notification related core imports from the CloudWatch `CloudWatchRequest`

This PR partially removes the following core imports, remaining imports will be removed in separate PRs:
```ts
import { notifyApp } from 'app/core/actions';
import { createErrorNotification } from 'app/core/copy/appNotification';
import { store } from 'app/store/store';
import { AppNotificationTimeout } from 'app/types';
```

**Why do we need this feature?**

This work is part of the work towards decoupling CloudWatch from core Grafana

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer:**

There's an ephemeral instance on HG that you can test this with. There's already a dashboard where I've added some template variables that can be tested. If you can't access it, let me know and I can share the dashboard JSON.

This PR swaps the core imports to use appEvents exported from `@grafana/runtime` to display app notifications.

The existing notification **_messages_** should be the same, but **_when_** they get triggered is different. Inline error messages are also shown for the invalid fields. Currently, the error message is always triggered if a variable is configured as a multi-enabled variable (a variable where you can select multiple options at once). This displays an error message even if the query is valid (for instance, if only a single option is selected). This PR makes it so the error message is only displayed if multiple options are selected and used somewhere that doesn't support multiple options being selected at once.

**Annotation using a dimension key that is a multi-variable and selecting multiple options vs selecting a single option**

https://github.com/grafana/grafana/assets/19530599/d6a3a8fb-ac2c-4bd1-8321-9aec69a5d3d2

**Variable query using a filter key with multiple options selected shows an inline error message**

https://github.com/grafana/grafana/assets/19530599/60b20bab-06fd-4872-b219-81a4581730e0

**Metric Query that uses a dimension key and dimension value that have multiple options selected and a single option selected. There are inline error messages to indicate where the invalid template variable is being used.**

https://github.com/grafana/grafana/assets/19530599/089b98e9-b1e9-4885-819e-b13fa689d56b

**Metric Query using a template variable that has multiple options selected generates a valid expression (see the `sqlExpression` generated in the payload of the query request. `$dimension_key` is interpolated to `InstanceId, ImageId` which is valid syntax in `SCHEMA` and `GROUP BY`)**

<img width="1274" alt="cloudwatchrequest - metric query code mode" src="https://github.com/grafana/grafana/assets/19530599/7d75fe15-7cd4-4198-bb7c-f1a6bbe654f0">

**Metric Search query using a multi-enabled dimension key with multiple options selected and then only selecting one option**

https://github.com/grafana/grafana/assets/19530599/9b41a8a8-7d0f-4a8b-a966-51429f1c776f

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

kevinwcyu committed:
CloudWatch: Remove core imports from CloudWatchRequest and use appEvents for notifications

kevinwcyu commented:
/deploy-to-hg

kevinwcyu committed:
add error message to variable query editor multi filters

kevinwcyu commented:
/deploy-to-hg

kevinwcyu committed:
fix tests

kevinwcyu committed:
fix lint

kevinwcyu committed:
fix lint

kevinwcyu commented on public/app/plugins/datasource/cloudwatch/query-runner/CloudWatchRequest.ts:
```diff
@@ -93,23 +95,35 @@ export abstract class CloudWatchRequest {
     }, {});
   }
 
+  isMultiVariable(target?: string) {
```
Logic taken from `replaceVariableAndDisplayWarningIfMulti`

kevinwcyu commented on public/app/plugins/datasource/cloudwatch/language/cloudwatch-sql/completion/CompletionItemProvider.ts:
```diff
@@ -179,12 +179,15 @@ export class SQLCompletionItemProvider extends CompletionItemProvider {
+                  metricName: metricNameToken?.value,
+                  dimensionFilters,
+                },
+                false
```
This disabled the warning about multi-template variables since Metric Search queries support a comma separated list of arguments for `FROM SCHEMA` and `GROUP BY`.

> If you specify SCHEMA with additional arguments, the additional arguments after the namespace argument must be label keys. Label keys must be dimension names. If you specify one or more of these label keys, the results are scoped down to only those metrics that have that exact set of dimensions. The order of these label keys does not matter.

> Groups the query results into multiple time series, each one corresponding to a different value for the specified label key or keys. For example, using GROUP BY InstanceId returns a different time series for each value of InstanceId. Using GROUP BY ServiceName, Operation creates a different time series for each possible combination of the values of ServiceName and Operation.

From https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch-metrics-insights-querylanguage.html

kevinwcyu commented on public/app/plugins/datasource/cloudwatch/query-runner/CloudWatchRequest.ts:
```diff
@@ -93,23 +95,35 @@ export abstract class CloudWatchRequest {
-          `Multi template variables are not supported for ${fieldName || target}`
-        );
-      }
+    if (displayErrorIfIsMultiTemplateVariable && this.isVariableWithMultipleOptionsSelected(target)) {
```
This now checks if a template variable is both a multi-enabled variable and if multiple options have been selected for it.

iwysiu commented on public/app/plugins/datasource/cloudwatch/hooks.test.ts:
```diff
@@ -139,4 +148,26 @@ describe('hooks', () => {
     });
   });
+
+  describe('useEnsureVariableHasSingleSelection', () => {
```
optional: probably trivial, but maybe a test case with a non-multi variable?

iwysiu reviewed (commented):
Cool! I tried it out and it worked well, though I have 2 notes about edge cases:
- I made 1 query for dimension keys ($keys) and 1 for dimension values ($values) and I used $keys for the dimension keys field of $values. When I was on the main dashboard, I switched $keys from 1 variable to 2, and $values quietly failed and got set to none. I feel like we should have some kind of alert or error pop up in this case?
- If you have multiple dimension filters the error popup makes the formatting a bit weird: 
![Screenshot 2024-01-31 at 4 17 18 PM](https://github.com/grafana/grafana/assets/5421859/ca6231de-2bbf-4e50-919c-7edb6d022f70)

kevinwcyu committed:
add test for non multi-variable

kevinwcyu committed:
pr comments

kevinwcyu committed:
Merge branch 'main' into kevinwcyu/remove-core-imports-from-cloudwatch-request

kevinwcyu commented:
> Cool! I tried it out and it worked well, though I have 2 notes about edge cases:
> 
> * I made 1 query for dimension keys ($keys) and 1 for dimension values ($values) and I used $keys for the dimension keys field of $values. When I was on the main dashboard, I switched $keys from 1 variable to 2, and $values quietly failed and got set to none. I feel like we should have some kind of alert or error pop up in this case?

I added an alert for this. The second part about $values silently failing and getting set to none is also how it currently behaves so I'd prefer to open up a separate ticket for that behaviour if that sounds good to you.

> * If you have multiple dimension filters the error popup makes the formatting a bit weird:
>   ![Screenshot 2024-01-31 at 4 17 18 PM](https://private-user-images.githubusercontent.com/5421859/301311162-ca6231de-2bbf-4e50-919c-7edb6d022f70.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY4MTcyNDIsIm5iZiI6MTcwNjgxNjk0MiwicGF0aCI6Ii81NDIxODU5LzMwMTMxMTE2Mi1jYTYyMzFkZS0yYmJmLTRlNTAtOTE5Yy03ZWRiNmQwMjJmNzAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMDFUMTk0OTAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWZmMzUxNjBmZDBkOTE0ZTQ2ODM5MDE2NTYxZDgxYTVmNjUzZTk3NDQ0YjExMzEzZDBlODliZDNjZDI2OWQxYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.XCVjessw9eTUXLH_fROo84W7ymx461RPRbm-mznGn7c)

I've added some constraints around the width of the alert messages in the areas where they are longer than the filters.

Changes for these are in this commit https://github.com/grafana/grafana/pull/81422/commits/e10c3171f9df569c4a24e448903f46ceb85f4187

idastambuk commented:
Nice! I enabled awsDatasourcesNewFormStyling toggle and it looks like the variable editor message should be wider than the form. Also even with the toggle disabled the query editor/annotations editor message isn't as wide as in the screen recordings - is it something about my setup? 

<img width="989" alt="Screenshot 2024-02-05 at 16 23 27" src="https://github.com/grafana/grafana/assets/16140639/a98ff919-6b16-4fa5-be97-e88831196bb2">
<img width="449" alt="Screenshot 2024-02-05 at 16 22 54" src="https://github.com/grafana/grafana/assets/16140639/ee6fee61-a9eb-4104-b550-f0a94d1530f4">

iwysiu reviewed (approved):
>I added an alert for this. The second part about $values silently failing and getting set to none is also how it currently behaves so I'd prefer to open up a separate ticket for that behaviour if that sounds good to you.

Sounds good!
I also confirmed that the alert size was updated

kevinwcyu committed:
Merge branch 'main' into kevinwcyu/remove-core-imports-from-cloudwatch-request
